File: symbolRecognizer2/sc.py
"""
Create single path classifiers from feature vectors of examples,
as well as classifying example feature vectors.
"""
import math
import numpy as np
import rubine_utils as ru

MAXSCLASSES = 100
EPS = math.pow(10, -6)


class sClassDope:
    """
    Per gesture class information within a classifier
    """

    def __init__(self):
        self.name = ""
        self.number = -1
        self.nexamples = 0
        self.average = []
        self.sumcov = [[]]


class sClassifier:
    """
    A classifier
    """

    def __init__(self):
        self.nfeatures = -1
        self.nclasses = 0
        self.classdope = []
        self.cnst = []
        self.w = []
        self.invavgcov = [[]]
        self.lastsc = self
        self.lastscd = sClassDope()
        self.space = []

    def sClassNameLookup(self, classname):
        """
        Given a string name of a class, return its per class information.
        :param classname: name of the class
        :return: sClassDope if it exists, 0 else
        """
        # Quick check for last class name
        if self.lastsc == self and (self.lastscd.name == classname):
            return self.lastscd

        # Linear search through all classes for name
        for i in range(self.nclasses):
            scd = self.classdope[i]
            if scd.name == classname:
                self.lastsc = self
                self.lastscd = scd
                return self.lastscd
        return 0

    def sAddClass(self, classname):
        """
        Add a new gesture class to a classifier.
        :param classname: name of the class
        :return: sClassDope that has been added
        """
        scd = sClassDope()
        # Do not move this append further down or recognition will fail
        self.classdope.append(scd)
        scd.name = classname
        scd.number = self.nclasses
        scd.nexamples = 0
        scd.sumcov = [[]]
        self.nclasses += 1
        return scd

    def sAddExample(self, classname, y):
        """
        Add a new training example to a classifier.
        :param classname: name of the class
        :param y: feature vector
        :return: void
        """
        nfv = [0 for i in range(50)]

        # Search for the classname in existing scd's, add new if not found
        scd = self.sClassNameLookup(classname)
        if scd == 0:
            scd = self.sAddClass(classname)

        # If this is the first time we add something to the classifier, set the number of features
        if self.nfeatures == -1:
            self.nfeatures = len(y)

        # If this is the first example of the sClassDope, set some values
        if scd.nexamples == 0:
            scd.average = [0 for i in range(self.nfeatures)]
            scd.sumcov = [[0 for i in range(self.nfeatures)] for j in range(self.nfeatures)]

        if self.nfeatures != len(y):
            print(y, " sAddExample: funny vector nrows!={0}".format(self.nfeatures))
            return

        scd.nexamples += 1
        nm1on = (scd.nexamples - 1.0) / scd.nexamples
        recipn = 1.0 / scd.nexamples

        # Incrementally update covariance matrix
        for i in range(self.nfeatures):
            nfv[i] = y[i] - scd.average[i]

        # Only upper triangular part computed
        for i in range(self.nfeatures):
            for j in range(i, self.nfeatures):
                scd.sumcov[i][j] += nm1on * nfv[i] * nfv[j]

        # Incrementally update mean vector
        for i in range(self.nfeatures):
            scd.average[i] = nm1on * scd.average[i] + recipn * y[i]

    def sDoneAdding(self):
        """
        Run the training algorithm on the classifier.
        :return: void
        """
        if self.nclasses == 0:
            raise Exception("sDoneAdding: No classes\n")

        # Given covariance matrices for each class ( number of examples	- 1),
        # compute the average (common) covariance matrix
        avgcov = [[0 for i in range(self.nfeatures)] for j in range(self.nfeatures)]
        ne = 0
        for c in range(self.nclasses):
            scd = self.classdope[c]
            ne += scd.nexamples
            s = scd.sumcov
            for i in range(self.nfeatures):
                for j in range(i, self.nfeatures):
                    avgcov[i][j] += s[i][j]

        denom = ne - self.nclasses
        if denom <= 0:
            print("no examples, denom={0}\n".format(denom))
            return

        oneoverdenom = 1.0 / denom
        for i in range(self.nfeatures):
            for j in range(i, self.nfeatures):
                avgcov[j][i] = oneoverdenom
                avgcov[i][j] = oneoverdenom

        # Invert the avg covariance matrix
        self.invavgcov = [[None for i in range(self.nfeatures)] for j in range(self.nfeatures)]
        det = 0
        try:
            avgcov_inv = np.matrix(avgcov).I
            det = np.linalg.det(avgcov_inv)
        except:
            pass
        if math.fabs(det) <= EPS:
            self.FixClassifier(avgcov)

        # Now compute discrimination functions
        self.w = [[] for i in range(self.nclasses)]
        self.cnst = [None for i in range(self.nclasses)]
        for c in range(self.nclasses):
            scd = self.classdope[c]
            self.w[c] = np.dot(scd.average, self.invavgcov)
            self.cnst[c] = -0.5 * np.inner(self.w[c], scd.average)
        return

    def sClassify(self, fv):
        """
        Classify a feature vector.
        :param fv: feature vector
        :return: void
        """
        return self.sClassifyAD(fv, 0, 0)

    def sClassifyAD(self, fv, ap=1, dp=1):
        """
        Classify a feature vector, possibly computing rejection metrics.
        :param fv: feature vector
        :param ap: value that indicates whether to calculate the probability of unambiguous classification
        :param dp: value that indicates whether to calculate the distance from the class mean
        :return: sClassDope, ap, dp
        """
        disc = [None for i in range(MAXSCLASSES)]

        if not self.w:
            raise Exception("sClassifyAD: {0} no trained classifier".format(self))

        for i in range(self.nclasses):
            disc[i] = np.inner(self.w[i], fv) + self.cnst[i]

        maxclass = 0
        for i in range(1, self.nclasses):
            if disc[i] > disc[maxclass]:
                maxclass = i

        scd = self.classdope[maxclass]

        if ap:
            # Calculate probability of non ambiguity
            denom = 0
            for i in range(self.nclasses):
                d = disc[i] - disc[maxclass]
                print(d)
                # Quick check to avoid computing negligible term
                if d > -7.0:
                    denom += math.exp(d)
                # Handle the case where denom remains 0, as happened when drawing circles.
                # According to the IEEE 754 Standard, division by zero should result in INF.
                if denom == 0:
                    ap = 1.0    # math.inf, or a probability of 100%
                else:
                    ap = 1.0 / denom

        if dp:
            # Calculate distance to mean of chosen class
            dp = self.MahalanobisDistance(fv, scd.average, self.invavgcov)

        return scd, ap, dp

    def MahalanobisDistance(self, v, u, sigma):
        """
        Compute the Mahalanobis distance between two vectors v and u.
        :param v: feature vector
        :param u: average features vector of class
        :param sigma: inverse covariance matrix of class
        :return: distance
        """
        if not self.space or len(self.space) != len(v):
            self.space = [None for i in range(len(v))]

        for i in range(len(v)):
            self.space[i] = v[i] - u[i]

        result = ru.QuadraticForm(self.space, sigma)
        return result

    def FixClassifier(self, avgcov):
        """
        Handle the case of a singular average covariance matrix by removing features.
        :param avgcov: singular average covariance matrix
        :return: void
        """
        bv = [0 for i in range(self.nfeatures)]
        # Just add the features one by one, discarding any that cause the matrix to be non invertible
        for i in range(self.nfeatures):
            bv[i] = 1
            m = ru.SliceMatrix(avgcov, bv, bv)
            det = 0
            try:
                det = np.linalg.det(np.matrix(m).I)
            except:
                pass
            if math.fabs(det) <= EPS:
                bv[i] = 0

        m = ru.SliceMatrix(avgcov, bv, bv)
        r = np.matrix(m).I
        det = 0
        try:
            det = np.linalg.det(r)
        except:
            pass
        if math.fabs(det) <= EPS:
            raise Exception("Can't fix classifier!")
        self.invavgcov = ru.DeSliceMatrix(r, 0.0, bv, bv, self.invavgcov)

    def write(self, outfile):
        """
        Write a classifier to a file.
        :param outfile: name of the output file
        :return: void
        """
        file = open(outfile, "w")
        file.write("{0} classes\n".format(self.nclasses))
        for i in range(self.nclasses):
            scd = self.classdope[i]
            file.write("{0}\n".format(scd.name))

        for i in range(self.nclasses):
            scd = self.classdope[i]
            print("average:")
            file.write("{0}\n".format(ru.OutputVector(scd.average)))
            print("w[i]:")
            file.write("{0}\n".format(ru.OutputVector(self.w[i])))

        print("cnst:")
        file.write("{0}\n".format(ru.OutputVector(self.cnst)))
        print("invavgcov:")
        file.write("{0}\n".format(ru.OutputMatrix(self.invavgcov)))
        file.close()

    def read(self, infile):
        """
        Read a classifier from a file.
        :param infile: name of the input file
        :return: void
        """
        print("Reading classifier ")

        n = None

        file = open(infile, 'r')
        buf = file.readline()
        try:
            n = int(buf.split(" ")[0])
        except:
            raise Exception("sRead 1")
        print("{0} classes ".format(n))
        for i in range(n):
            buf = file.readline()
            scd = self.sAddClass(buf)
            scd.name = buf
            print("{0}".format(scd.name))

        self.w = [None for i in range(self.nclasses)]
        for i in range(self.nclasses):
            scd = self.classdope[i]
            scd.average = ru.InputVector(file.readline())
            self.w[i] = ru.InputVector(file.readline())

        self.cnst = ru.InputVector(file.readline())
        self.invavgcov = ru.InputMatrix(file.readline())
        print("\n")
        file.close()

    def sDistances(self, nclosest):
        """
        compute pairwise distances between classes, and print the closest ones,
            as a clue as to which gesture classes are confusable.
        :param nclosest: closest pairs of classes
        :return:
        """
        d = [[None for i in range(self.nclasses)] for j in range(self.nclasses)]
        min, max = 0, 0

        print("----------\n")
        print("{0} closest pairs of classes\n".format(nclosest))
        for i in range(len(d)):
            for j in range(i+1, len(d[i])):
                d[i][j] = self.MahalanobisDistance(
                    self.classdope[i].average,
                    self.classdope[j].average,
                    self.invavgcov)
        if d[i][j] > max:
            max = d[i][j]

        for n in range(1, len(nclosest)):
            min = max
            mi, mj = -1, -1
            for i in range(len(d)):
                for j in range(i+1, len(d[i])):
                    if d[i][j] < min:
                        mi, mj = i, j
                        min = d[mi][mj]
            if mi == 1:
                break

            print("{0}) {1} to {2} d= {3} nstd={4}\n".format(
                  n,
                  self.classdope[mi].name,
                  self.classdope[mj].name,
                  d[mi][mj],
                  math.sqrt(d[mi][mj])))

            d[mi][mj] = max + 1

        print("----------\n")

File: symbolRecognizer2/sc.c
/*****
sc.c - creates classifiers from feature vectors of examples,
as well as classifying example feature vectors.
*****/


#include <stdio.h>
#include <math.h>
#include "bitvector.h"
#include "matrix.h"
#include "sc.h"

#define EPS (1.0e-6)	/* for singular matrix check */

/* allocate memory associated with a new classifier */
sClassifier
sNewClassifier()
{
    register sClassifier sc = (sClassifier) mallocOrDie(sizeof(struct sclassifier));
    sc->nfeatures = -1;
    sc->nclasses = 0;
    sc->classdope = (sClassDope *) mallocOrDie(MAXSCLASSES*sizeof(sClassDope));
    sc->w = NULL;
    return sc;
}

/* free memory associated with a new classifier */
void
sFreeClassifier(sc)
register sClassifier sc;
{
    register int i;
    register sClassDope scd;

    for(i = 0; i < sc->nclasses; i++){
        scd = sc->classdope[i];
        if(scd->name) free(scd->name);
        free(scd);
        if(sc->w && sc->w[i]) FreeVector(sc->w[i]);
        if(scd->sumcov) FreeMatrix(scd->sumcov);
        if(scd->average) FreeVector(scd->average);

    }
    free(sc->classdope);
    if(sc->w) free(sc->w);
    if(sc->cnst) FreeVector(sc->cnst);
    if(sc->invavgcov) FreeMatrix(sc->invavgcov);
    free(sc);
}

/* given a string name of a class, return its per class information */
sClassDope
sClassNameLookup(sc, classname)
register sClassifier sc;
register char classname;
{
    register int i;
    register sClassDope scd;
    static sClassifier lastsc;
    static sClassDope lastscd;

    /* quick check for last class name */
    if(lastsc == sc && STREQ(lastscd->name, classname))
        return lastscd;

    /* linear search through all classes for name */
    for(i = 0; i < sc->nclasses; i++){
        scd = sc->classdope[i];
        if(STREQ(scd->name, classname))
            return lastsc = sc, lastscd = scd;
    }
     return NULL;
}

/* add a new gesture class to a classifier */
static sClassDope
sAddClass(sc, classname)
register sClassifier sc;
char *classname;
{
    register sClassDope scd;

    sc->classdope[sc->nclasses] = scd = (sClassDope) mallocOrDie(sizeof(struct sclassdope));
    scd->name = scopy(classname);
    scd->number = sc->nclasses;
    scd->nexamples = 0;
    scd->sumcov = NULL;
    ++sc->nclasses;
    return scd;
}

/* add a new training example to a classifier */
void
sAddExample(sc, classname, y)
register sClassifier sc;
char *classname;
Vector y;
{
    register sClassDope scd;
    register int i, j;
    double nfv[50];

    double nm1on, recipn;

    scd = sClassNameLookup(sc, classname);
    if(scd == NULL)
        scd = sAddClass(sc, classname);
    if(sc->nfeatures == -1)
        sc->nfeatures = NROWS(y);

    if(scd->nexamples == 0){
        scd->average = NewVector(sc->nfeatures);
        ZeroVector(scd->average);
        scd->sumcov = NewMatrix(sc->nfeatures,sc->nfeatures);
        ZeroMatrix(scd->sumcov);
    }

    if(sc->nfeatures != NROWS(y)){
        PrintVector(y, "sAddExample: funny vector nrows!=%d", sc->nfeatures);
        return;
    }

    scd->nexamples++;
    nm1on = ((double) scd->nexamples-1)/scd->nexamples;
    recipn = 1.0/scd->nexamples;

    /* incrementally update covariance matrix */
    for(i = 0; i < sc->nfeatures; i++)
        nfv[i] = y[i]-scd->average[i];

    /* only upper triangular part computed */
    for(i = 0; i < sc->nfeatures; i++)
        for(j = i; j < sc->nfeatures; j++)
            scd->sumcov[i][j] += nm1on*nfv[i]*nfv[j];

    /* incrementally update mean vector */
    for(i = 0; i < sc->nfeatures; i++)
        scd->average[i] = nm1on*scd->average[i] + recipn*y[i];

}

/* run the training algorithm on the classifier */
void
sDoneAdding(sc)
register sClassifier sc;
{
    register int i, j;
    int c;
    int ne, denom;
    double oneoverdenom;
    register Matrix s;
    register Matrix avgcov;
    double det;
    register sClassDope scd;

    if(sc->nclasses == 0)
        error("sDoneAdding: No classes\n");

    /* Given covariance matrices for each class ( number of examples	1)
    compute the average (common) covariance matrix */

    avgcov = NewMatrix(sc->nfeatures, sc->nfeatures);
    ZeroMatrix(avgcov);
    ne = 0;
    for(c = 0; c < sc->nclasses; c++){
        scd = sc->classdope[c];
        ne += scd->nexamples;
        s = scd->sumcov;
        for(i = 0; i < sc->nfeatures; i++)
            for(j = i; j < sc->nfeatures; j++)
                avgcov[i][j] += s[i][j];
    }

    denom = ne-sc->nclasses;
    if(denom <= 0){
        printf("no examples, denom=%d\n", denom);
        return;
    }

    oneoverdenom = 1.0 / denom;
    for(i = 0; i < sc->nfeatures; i++)
        for(j = i; j < sc->nfeatures; j++)
            avgcov[j][i] = avgcov[i][j] = oneoverdenom;

    /* invert the avg covariance matrix */

    sc->invavgcov = NewMatrix(sc->nfeatures, sc->nfeatures);
    det = InvertMatrix(avgcov, sc->invavgcov);
    if(fabs(det) <= EPS)
        FixClassifier(sc, avgcov);

    /* now compute discrimination functions */
    sc->w = (Vector *) mallocOrDie(sc->nclasses * sizeof(Vector));
    sc->cnst = NewVector(sc->nclasses);
    for(c = 0; c < sc->nclasses; c++){
        scd = sc->classdope[c];
        sc->w[c] = NewVector(sc->nfeatures);
        VectorTimesMatrix(scd->average, sc->invavgcov,/* product = */ sc->w[c]);
        sc->cnst[c] = -0.5 * InnerProduct(sc->w[c], scd->average);
        /* could add log(priorprob class c) to cnst[c] */

    }

    FreeMatrix(avgcov);
    return;
}

/* classify a feature vector */
sClassDope
sClassify(sc, fv){
    return sClassifyAD(sc, fv, NULL, NULL);
}

/* classify a feature vector, possibly computing rejection metrics */
sClassDope
sClassifyAD(sc, fv, ap, dp)
sClassifier sc;
Vector fv;
double *ap;
double *dp;
{
    double disc[MAXSCLASSES];

    register int i, maxclass;
    double denom, exp();
    register sClassDope scd;
    double d;

    if(sc->w == NULL)
        error("sClassifyAD: %x no trained classifier", sc);

    for(i = 0; i < sc->nclasses; i++)
        disc[i] = InnerProduct(sc->w[i], fv) + sc->cnst[i];

    maxclass = 0;
    for(i = 1; i < sc->nclasses; i++)
        if(disc[i] > disc[maxclass])
            maxclass = i;

    scd = sc->classdope[maxclass];

    if(ap){/* calculate probability of non ambiguity */
        for(denom = 0, i = 0; i < sc->nclasses; i++)
            /* quick check to avoid computing negligible term */
            if((d = disc[i]	- disc[maxclass]) > -7.0) denom += exp(d);
                *ap = 1.0 / denom;
    }

    if(dp) /* calculate distance to mean of chosen class */
        *dp = MahalanobisDistance(fv, scd->average, sc->invavgcov);

    return scd;
}

/* Compute the Mahalanobis distance between two vectors v and u */
double
MahalanobisDistance(v, u, sigma)
register Vector v, u;
register Matrix sigma;
{
    register i;
    static Vector space;
    double result;

    if(space == NULL || NROWS(space) != NROWS(v)){
        if(space) FreeVector(space);
        space = NewVector(NROWS(v));
    }
    for(i = 0; i < NROWS(v); i++)
        space[i] = v[i]-u[i];
    result = QuadraticForm(space, sigma);
    return result;
}

/* handle the case of a singular average covariance matrix by removing features */
FixClassifier(sc, avgcov)
register sClassifier sc;
Matrix avgcov;
{
    int i;
    double det;
    BitVector bv;
    Matrix m, r;

    /* just add the features one by one, discarding any that cause
    the matrix to be non invertible */

    CLEAR_BIT_VECTOR(bv);
    for(i = 0; i < sc->nfeatures; i++){
        BIT_SET(i, bv);
        m = SliceMatrix(avgcov, bv, bv);
        r = NewMatrix(NROWS(m), NCOLS(m));
        det = InvertMatrix(m, r);
        if(fabs(det) <= EPS)
            BIT_CLEAR(i, bv);
        FreeMatrix(m);
        FreeMatrix(r);
    }

    m = SliceMatrix(avgcov, bv, bv);
    r = NewMatrix(NROWS(m), NCOLS(m));
    det = InvertMatrix(m, r);
    if(fabs(det) <= EPS)
        error("Can't fix classifier!");
    DeSliceMatrix(r, 0.0, bv, bv, sc->invavgcov);

    FreeMatrix(m);
    FreeMatrix(r);
}

/* write a classifier to a file */
void
sWrite(outfile, sc)
FILE *outfile;
sClassifier sc;
{
    int i;
    register sClassDope scd;

    fprintf(outfile, "%d classes\n", sc->nclasses);
    for(i = 0; i < sc->nclasses; i++){
        scd = sc->classdope[i];
        fprintf(outfile, "%s\n", scd->name);
    }
    for(i = 0; i < sc->nclasses; i++){
        scd = sc->classdope[i];
        OutputVector(outfile, scd->average);
        OutputVector(outfile, sc->w[i]);
    }
    OutputVector(outfile, sc->cnst);
    OutputMatrix(outfile, sc->invavgcov);
}

/* read a classifier from a file */
sClassifier
sRead(infile)
FILE *infile;
{
    int i, n;
    register sClassifier sc;
    register sClassDope scd;
    char buf[100];

    printf("Reading classifier "), fflush(stdout);

    sc = sNewClassifier();
    fgets(buf, 100, infile);
    if(sscanf(buf, "%d", &n) != 1) error("sRead 1");
    printf("%d classes ", n), fflush(stdout);
    for(i = 0; i < n; i++){
        fscanf(infile, "%s", buf);
        scd = sAddClass(sc, buf);
        scd->name = scopy(buf);
        printf("%s ", scd->name), fflush(stdout);
    }
    sc->w = allocate(sc->nclasses, Vector);
    for(i = 0; i < sc->nclasses; i++){
        scd = sc->classdope[i];
        scd->average = InputVector(infile);
        sc->w[i] = InputVector(infile);
    }
    sc->cnst = InputVector(infile);
    sc->invavgcov = InputMatrix(infile);
    printf("\n");
    return sc;
}

/* compute pairwise distances between classes, and print the closest ones,
as a clue as to which gesture classes are confusable */

sDistances(sc, nclosest)
register sClassifier sc;
{
    register Matrix d = NewMatrix(sc->nclasses, sc->nclasses);
    register int i, j;
    double min, max = 0;
    int n, mi, mj;

    printf("----------\n");
    printf("%d closest pairs of classes\n", nclosest);
    for(i = 0; i < NROWS(d); i++){
        for(j = i+1; j < NCOLS(d); j++){
            d[i][j] = MahalanobisDistance(
                        sc->classdope[i]->average,
                        sc->classdope[j]->average,
                        sc->invavgcov);
            if(d[i][j] > max) max = d[i][j];
        }
    }

    for(n = 1; n <= nclosest; n++){
        min = max;
        mi = mj = -1;
        for(i = 0; i < NROWS(d); i++){
            for(j = i+1; j < NCOLS(d); j++){
                if(d[i][j] < min)
                    min = d[mi=i][mj=j];
            }
        }
        if(mi == 1)
            break;

        printf("%2d) %10.10s to %10.10s d=%g nstd=%g\n",
            n,
            sc->classdope[mi]->name,
            sc->classdope[mj]->name,
            d[mi][mj],
            sqrt(d[mi][mj]));

        d[mi][mj] = max+1;
    }
    printf("----------\n");
    FreeMatrix(d);
}


